### ISSUE TITLE
[MDEPLOY-176] With DeployAtEnd activated and a project using customized lifecycle plugin, deploy is skipped 

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-deploy-plugin#354]
**[olivier o](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olivier%20o.)** opened **[MDEPLOY-176](https://issues.apache.org/jira/browse/MDEPLOY-176?redirect=false)** and commented

Hi,
When deployAtEnd is enabled, the deployment of all modules is skipped. It seems that the "readyProjectsCounter" is not accurate for one module. So the last module is not detected.

I've attached a simple sample project with 4 modules :
- pom : set version and plugin management
- jar and ear  : empty project (only to illustrate counter reset)
- war : module with readyProjectsCounter

I've looked at 2.9-SNAP version code , and displayed some info from attached sample :
pom : readyProjectsCounter: 1 - reactorProjects: 4
jar: readyProjectsCounter: 2 - reactorProjects: 4
war: **readyProjectsCounter: 1** - reactorProjects: 4
ear: readyProjectsCounter: 3 - reactorProjects: 4

Regards,
Olivier

---

**Affects:** 2.8.1, 2.8.2

**Attachments:**
- [deployAtEndFailure.zip](https://issues.apache.org/jira/secure/attachment/12764421/deployAtEndFailure.zip) (_10.56 kB_)
- [deploy-end-fail.zip](https://issues.apache.org/jira/secure/attachment/12715241/deploy-end-fail.zip) (_225.19 kB_)

**Issue Links:**
- [MDEPLOY-157](https://issues.apache.org/jira/browse/MDEPLOY-157) Add deployAtEnd option for multimodule projects

- [MDEPLOY-201](https://issues.apache.org/jira/browse/MDEPLOY-201) deployAtEnd issue for multi-module projects

- [MDEPLOY-193](https://issues.apache.org/jira/browse/MDEPLOY-193) deployAtEnd does not deploy artifacts for multi-module project with build extensions

6 votes, 13 watchers

**Comment by jira-importer** [comment]
**[Jörg von Frantzius](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jfrantzius)** commented

Hi Olivier,

the title of this issue mentions "project using customized lifecycle plugin", could you explain what you mean by this? I looked at the sample and couldn't find anything like that. If this is outdated then maybe you can remove this from your issue title?

By the way, I can reproduce the problem with your sample and maven-deploy-plugin 2.8.1 and a simple "mvn deploy", and I have a similar problem with my own project.

**Comment by jira-importer** [comment]
**[olivier o](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olivier%20o.)** commented

Hi JÃ´rg,
Here some details, in fact in the war module \<artifactId>deploy-end-fail-war\</artifactId>, I use  this plugin
\<groupId>com.devspan.mojo.javascript\</groupId>
\<artifactId>javascript-maven-plugin\</artifactId>
\<version>0.9.3\</version>
It seems this plugin customize the maven lifecycle, and I think if you comment the plugin, it's working.
Is it clearer ?

**Comment by jira-importer** [comment]
**[Christian Lipphardt (camunda)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hawky4s)** commented

Hi all,

I can reproduce the problem with the org.apache.felix:maven-bundle-plugin:2.4.0 in a simple parent pom + jar multi module build when using the \<extensions>true\</extensions> element for the maven-bundle-plugin as required. When I comment the \<extensions> element, the deployAtEnd will be executed, otherwise it will just be skipped.

**Comment by jira-importer** [comment]
**[bob](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bob_h)** commented

Hi all, I have a similar problem.
As I am migrating from maven 2, I can precise that the deployAtEnd feature is working with maven 2.2.1 but no more with maven 3.2.5 and 3.3.3.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

So you are using a customized life cycle ? link bundle-plugin or something similar? I'm working on Maven 3.1.1 and works perfect...

**Comment by jira-importer** [comment]
**[bob](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bob_h)** commented

Hi Karl, I used the maven bundle plugin and removed it to be able to use the deployAtEnd.
I still have the issue for building a project with a war.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

The questions is what exactly the issue is with a project with a war file? Can create an example project which you put on github or something similar so we have something concrete to talk about...

**Comment by jira-importer** [comment]
**[bob](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bob_h)** commented

I have a multi module project. One of these modules is using the war packaging.

For the deployment I set the deployAtEnd flag to true.
If I use maven 2.2.1,
the deploy is working.

If I use maven 3.2.5 or 3.1.1,
we have the log Deploying XXX at end but at the end nothing is uploaded.
If we remove the war from the list of the modules the artifacts are uploaded.
If I only try to deploy the war (directly from the "submodule") it is working.

Peharps the discussion is more about MDEPLOY-201 or MDEPLOY-193.

As I had the ( OSGi bundle and war use case in a multi module project), I could build a sample.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

You wrote your deploy is working with Maven 2.2.1 Ok..first Maven 2.2.1 does not support parallel builds and of course do you use parallel builds like `mvn -T 4 clean deploy` ?...Furthermore does your build work with an empty local cache which means a simple `mvn clean package` will it work ? Apart from that what version of `maven-war-plugin` do you use? And as a asked before can you make an example project which shows the wrong behaviour ? Ah..one more thing do all modules have correct parent/child relation ship ?

**Comment by jira-importer** [comment]
**[bob](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bob_h)** commented

I found the issue with the war,
this pom was redefining the version of maven deploy (2.8.1 instead of 2.8.2)
So there is no issue with the WAR use case.

I attached a project example which is failing with a bundle packaging. The jar will be empty but should be enough for your test.
I put multiple profiles to be able to build jar / war and osgi modules.
Run the command
mvn clean deploy -Pjar,war,osgi
